sycl/include/sycl/ext/oneapi/experimental/builtins.hpp

```diff
@@ -115,6 +142,31 @@ inline __SYCL_ALWAYS_INLINE
 #endif
 }
 
+template <size_t N>
+inline __SYCL_ALWAYS_INLINE sycl::marray<half, N>
+exp2(sycl::marray<half, N> x) __NOEXC {
+  sycl::marray<half, N> res;
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
+  for (size_t i = 0; i < N / 2; i++) {
+    auto partial_res = __clc_native_exp2(sycl::detail::to_vec(x, i * 2));
+    std::memcpy(&res[i * 2], &partial_res, sizeof(vec<half, 2>));
+  }
+  if constexpr (N % 2) {
+    res[N - 1] = __clc_native_exp2(x[N - 1]);
+  }
+#else
+  for (size_t i = 0; i < N / 2; i++) {
+    auto partial_res = __sycl_std::__invoke_exp2<sycl::vec<half, 2>>(
+        sycl::detail::to_vec(x, i * 2));
+    std::memcpy(&res[i * 2], &partial_res, sizeof(vec<half, 2>));
```

Comment:
Do you mean an explicit conversion operator allowing something like 
```
marray<T, N> x;
marray<vec<T, 2>, N/2> x_vec2 = x;
``` 
or similar.

Or an implicit conversion operator from `x[i * 2]` to `vec<T, 2>` etc?

I think that the main issue (beyond the question of whether this would make the code more readable or less confusing) is that this would have to break SYCL spec definitions of  marray/vec/etc? Perhaps I am missing something or misunderstanding what you mean?

Thanks

Reply:
marray is supposed to be more general than vec. 
What is the long term plan with vec? Will marray replace vec?
@Pennycook, can you comment on that?

Reply:
So this marray implementation supports any marray size N where N is in the set of size_t: so it is more general than vec in this sense. I think that it makes sense to map the marray implementation onto the vec one like e.g. I have done here by using vec2 (although it would be a fair argument to suggest another impl that makes use of larger vec sizes, as we've discussed in another thread). In theory this should also allow for vectorized register loads, although for some reason when we cast from marray to vec the vectorized loads that we see when using the standard vec implementation are not used: This is something we have as a TODO to improve on.

Reply:
There was a good presentation by your colleagues at syclcon.org this year that talks about it: *Untangling Modern Parallel Programming Models* https://www.youtube.com/watch?v=6FbW6zVYkxk&list=PL46sP9LM8GsyHAxj1k7MbWrv5f5SlMpIF&index=27

Reply:
> marray is supposed to be more general than vec. What is the long term plan with vec? Will marray replace vec? @Pennycook, can you comment on that?

+@rolandschulz

Reply:
We added `marray` in SYCL 2020 to untangle `sycl::vec`. The problem with `sycl::vec` is that it serves two different use case: explicit SIMD programming (vectorization across it with subgroup size of 1) and a simple math array (vectorization across the subgroup). And those two use cases require different API (e.g. you don't want element wise access for a SIMD vector that's why `sycl::vec` has swizzle). For the SIMD use-case, DPC++ has for now ESIMD. For SYCL 2023 we would like to add a new class aligned with `std::simd` and deprecate `sycl::vec`.

Reply:
> Do you mean an explicit conversion operator allowing something like
> 
> ```
> marray<T, N> x;
> marray<vec<T, 2>, N/2> x_vec2 = x;
> ```
> 
> or similar.
> 
> Or an implicit conversion operator allowing `x[i * 2]` -> `vec<T, 2>` etc?
> 
> I think that the main issue (beyond the question of whether this would make the code more readable or less confusing) is that this would have to break SYCL spec definitions of marray/vec/etc? Perhaps I am missing something or misunderstanding what you mean?
> 
> Thanks

@keryell do you have any thoughts on the above?

Thanks

Reply:
> Do you mean an explicit conversion operator allowing something like
> 
> ```
> marray<T, N> x;
> marray<vec<T, 2>, N/2> x_vec2 = x;
> ```
> 
> or similar.
> 
> Or an implicit conversion operator allowing `x[i * 2]` -> `vec<T, 2>` etc?
> 
> I think that the main issue (beyond the question of whether this would make the code more readable or less confusing) is that this would have to break SYCL spec definitions of marray/vec/etc? Perhaps I am missing something or misunderstanding what you mean?

Yes, this would be the idea.
I was just thinking here to make your extension code clearer, by having an implicit linearization of the recursive types.
Whether we expose this officially in SYCL is another story. But if enough people think is useful, why not?

Reply:
I didn't think it was really possible to have such conversions without exposing it in SYCL since I had thought C++ requires such conversion operators to be defined within the class, and if they could be defined outside of the class then it could be hard for someone to find the definition of the conversion operator. Conversions between `vec<T, 2>` and `marray<T, N>` or `T` look strange to me and I wouldn't know what it is doing without looking up the conversion operator.

Reply:
> I think that it makes sense to map the marray implementation onto the vec one like e.g. I have done here by using vec2 (although it would be a fair argument to suggest another impl that makes use of larger vec sizes, as we've discussed in another thread). In theory this should also allow for vectorized register loads, although for some reason in the CUDA backend when we cast from marray to vec the vectorized loads that we see when using the standard vec implementation are not used: This is something we have as a TODO to improve on.

We've discussed this a bit offline.
The reason why loads and stores to/from `marray`s are not being vectorized is that LLVM's `load-store-vectorizer` pass has a strict requirement on the alignment; the pointers have to be aligned to at least what the resulting vector would require (or the target must allow misaligned operations). The alignment requirement can be easily achieved by changing the default alignment of `marray` to the "previous" vector (i.e.: `marray<15, T>` would be decorated with `__attribute__((aligned(8 * sizeof(T)))`, making sure it's a power of 2 number).

WRT behind the scenes conversion of `marray` elements to vectors (`to_vec2`), it seems wrong, it is unlikely to bring performance benefits, as it will always result in temporary storage and extra loads, stores instructions. Perhaps we could use the same approach as cutlas does, harnessing the fact that alignment is set correctly, we could do a bit of type punning see: https://github.com/NVIDIA/cutlass/blob/e7a61c761a4bfb387b61c03cdbcd19ab300726b7/include/cutlass/functional.h#L1444

I had a go at the above in here: https://github.com/JackAKirk/llvm/pull/1

As a side note, it feels like somewhere here there is a compiler optimization missed, we should be able to use the same logic as the vectorizer pass and gather those scalar intrinsic, converting them to vector equivalents, making this code a lot cleaner.

Reply:
@aelovikov-intel @keryell 

What do you think about @jchlanda 's suggestion above?

Reply:
I don't think it becomes less UB.

> it seems wrong, it is unlikely to bring performance benefits, as it will always result in temporary storage and extra loads, stores instructions.

Why wouldn't the compiler optimize this? This is a standard C++ idiom for the compiler - https://en.cppreference.com/w/cpp/numeric/bit_cast

Reply:
> Why wouldn't the compiler optimize this? This is a standard C++ idiom for the compiler - https://en.cppreference.com/w/cpp/numeric/bit_cast

The compiler mustn't optimize it away because of the difference in the alignment of (elements of) `marray` and the temporary `vec2` variable, as `marray` follows the `std::array` alignment rules. For the same reason compiler was unable to generate vector loads and stores directly to/from `marray`, this is enforced in PTX spec:

> By default, vector variables are aligned to a multiple of their overall size (vector length times base-type size), to enable vector load and store instructions which require addresses aligned to a multiple of the access size.

This patch changes it though, and `marray` satisfies the alignment requirement of the "previous" vector. 

> I don't think it becomes less UB.

You are right, I reverted the type punning casts, with alignment fix LLVM is clever enough to optimize `vec2` variable and a call to `memcpy` away.

Reply:
> 
> > I don't think it becomes less UB.
> 
> You are right, I reverted the type punning casts, with alignment fix LLVM is clever enough to optimize `vec2` variable and a call to `memcpy` away.

Sounds like we have a winner.